### ISSUE TITLE
FailedMount - Error while creating pod with vSphere volume

### ISSUE THREAD

**Issue by BaluDontu** [vmware/kubernetes#231]
I am trying to use the latest master code base for deploying a kubernetes cluster. After successful deployment of kubernetes cluster, I tried to create a pod with dynamically provisioned vsphere volume and it errors out with the following reason:

FailedMount: MountVolume.MountDevice failed for volume "pvc-1bc80cc4-8214-11e7-848c-0050569c5070" : executable file not found in $PATH.

Because of this error, I see the volume is not mounted on the node. But the volume is successfully attached to the node.

```
Events:
  FirstSeen	LastSeen	Count	From			SubObjectPath	Type		Reason			Message
  ---------	--------	-----	----			-------------	--------	------			-------
  3m		3m		1	{default-scheduler }			Normal		Scheduled		Successfully assigned redis-slave-1850905298-fzbx2 to node3
  3m		3m		1	{kubelet node3}				Normal		SuccessfulMountVolume	MountVolume.SetUp succeeded for volume "default-token-4vxnn" 
  2m		1m		7	{kubelet node3}				Warning		FailedMount		MountVolume.MountDevice failed for volume "pvc-1bc80cc4-8214-11e7-848c-0050569c5070" : executable file not found in $PATH
```

@tusharnt  @divyenpatel  @luomiao @rohitjogvmw

**Comment by BaluDontu** [comment]
This is currently blocking my testing for https://github.com/vmware/kubernetes/issues/175.

**Comment by divyenpatel** [comment]
This issue is not observed on the stable master build  taken from https://storage.googleapis.com/kubernetes-release-dev/ci/latest-green.txt 
Build: v1.8.0-alpha.2.1117+5bb42cbc5888ef

Tests are able to attache volume in the nightly Jenkins Job for master branch.
http://cna-storage-jenkins.eng.vmware.com:8080/job/k8s-release-dev-ci-latest/187/'


@BaluDontu which executable file it is looking for mounting volume? Can we check from the log?

**Comment by BaluDontu** [comment]
I get the following errors in the kubelet.

> {"log":"I0816 20:56:15.900236     659 operation_generator.go:417] MountVolume.WaitForAttach succeeded for volume \"pvc-4dc51e93-82c5-11e7-a497-0050569c48bf\" (UniqueName: \"kubernetes.io/vsphere-volume/[vsanDatastore] kubevols/kubernetes-dynamic-pvc-4dc51e93-82c5-11e7-a497-0050569c48bf.vmdk\") pod \"thindiskpod\" (UID: \"52fbf7a2-82c5-11e7-a497-0050569c48bf\") \n","stream":"stderr","time":"2017-08-16T20:56:15.901072134Z"}

> {"log":"I0816 20:56:15.902508     659 mount_linux.go:413] Checking for issues with fsck on disk: /dev/disk/by-id/wwn-0x6000c29d3c96f77096950581483fe3a5\n","stream":"stderr","time":"2017-08-16T20:56:15.902695897Z"}

> {"log":"I0816 20:56:15.921616     659 mount_linux.go:427] `fsck` error fsck from util-linux 2.25.2\n","stream":"stderr","time":"2017-08-16T20:56:15.921725608Z"}

> {"log":"fsck: error 2 (No such file or directory) while executing fsck.ext2 for /dev/sdb\n","stream":"stderr","time":"2017-08-16T20:56:15.921752327Z"}

> {"log":"I0816 20:56:15.921652     659 mount_linux.go:432] Attempting to mount disk: ext4 /dev/disk/by-id/wwn-0x6000c29d3c96f77096950581483fe3a5 /var/lib/kubelet/plugins/kubernetes.io/vsphere-volume/mounts/[vsanDatastore] kubevols/kubernetes-dynamic-pvc-4dc51e93-82c5-11e7-a497-0050569c48bf.vmdk\n","stream":"stderr","time":"2017-08-16T20:56:15.921774238Z"}

> {"log":"I0816 20:56:15.921686     659 mount_linux.go:160] Mounting cmd (mount) with arguments ([-t ext4 -o defaults /dev/disk/by-id/wwn-0x6000c29d3c96f77096950581483fe3a5 /var/lib/kubelet/plugins/kubernetes.io/vsphere-volume/mounts/[vsanDatastore] kubevols/kubernetes-dynamic-pvc-4dc51e93-82c5-11e7-a497-0050569c48bf.vmdk])\n","stream":"stderr","time":"2017-08-16T20:56:15.921948786Z"}

> {"log":"E0816 20:56:15.929506     659 mount_linux.go:165] Mount failed: exit status 32\n","stream":"stderr","time":"2017-08-16T20:56:15.929619467Z"}

> {"log":"Mounting command: mount\n","stream":"stderr","time":"2017-08-16T20:56:15.92963496Z"}

> {"log":"Mounting arguments: -t ext4 -o defaults /dev/disk/by-id/wwn-0x6000c29d3c96f77096950581483fe3a5 /var/lib/kubelet/plugins/kubernetes.io/vsphere-volume/mounts/[vsanDatastore] kubevols/kubernetes-dynamic-pvc-4dc51e93-82c5-11e7-a497-0050569c48bf.vmdk\n","stream":"stderr","time":"2017-08-16T20:56:15.929639337Z"}

> {"log":"Output: mount: wrong fs type, bad option, bad superblock on /dev/sdb,\n","stream":"stderr","time":"2017-08-16T20:56:15.929642498Z"}

> {"log":"       missing codepage or helper program, or other error\n","stream":"stderr","time":"2017-08-16T20:56:15.929645312Z"}

> {"log":"\n","stream":"stderr","time":"2017-08-16T20:56:15.929647822Z"}

> {"log":"       In some cases useful info is found in syslog - try\n","stream":"stderr","time":"2017-08-16T20:56:15.929650562Z"}

> {"log":"       dmesg | tail or so.\n","stream":"stderr","time":"2017-08-16T20:56:15.929653095Z"}

> {"log":"\n","stream":"stderr","time":"2017-08-16T20:56:15.929655488Z"}

> {"log":"I0816 20:56:15.931335     659 mount_linux.go:480] Attempting to determine if disk \"/dev/disk/by-id/wwn-0x6000c29d3c96f77096950581483fe3a5\" is formatted using lsblk with args: ([-n -o FSTYPE /dev/disk/by-id/wwn-0x6000c29d3c96f77096950581483fe3a5])\n","stream":"stderr","time":"2017-08-16T20:56:15.931446377Z"}

> {"log":"I0816 20:56:15.938198     659 mount_linux.go:483] Output: \"\\n\"\n","stream":"stderr","time":"2017-08-16T20:56:15.938284852Z"}

> {"log":"I0816 20:56:15.938236     659 mount_linux.go:452] Disk \"/dev/disk/by-id/wwn-0x6000c29d3c96f77096950581483fe3a5\" appears to be unformatted, attempting to format as type: \"ext4\" with options: [-F /dev/disk/by-id/wwn-0x6000c29d3c96f77096950581483fe3a5]\n","stream":"stderr","time":"2017-08-16T20:56:15.938393307Z"}

> {"log":"E0816 20:56:15.939026     659 mount_linux.go:460] format of disk \"/dev/disk/by-id/wwn-0x6000c29d3c96f77096950581483fe3a5\" failed: type:(\"ext4\") target:(\"/var/lib/kubelet/plugins/kubernetes.io/vsphere-volume/mounts/[vsanDatastore] kubevols/kubernetes-dynamic-pvc-4dc51e93-82c5-11e7-a497-0050569c48bf.vmdk\") options:([\"defaults\"])error:(executable file not found in $PATH)\n","stream":"stderr","time":"2017-08-16T20:56:15.939126395Z"}

> {"log":"I0816 20:56:15.939915     659 server.go:214] Event(v1.ObjectReference{Kind:\"Pod\", Namespace:\"default\", Name:\"thindiskpod\", UID:\"52fbf7a2-82c5-11e7-a497-0050569c48bf\", APIVersion:\"v1\", ResourceVersion:\"5203\", FieldPath:\"\"}): type: 'Warning' reason: 'FailedMount' MountVolume.MountDevice failed for volume \"pvc-4dc51e93-82c5-11e7-a497-0050569c48bf\" : executable file not found in $PATH\n","stream":"stderr","time":"2017-08-16T20:56:15.94001794Z"}

I did the following steps to get the latest code from master.
> git fetch master
> git checkout -b k8smaster upstream/master
Used our jenkins build to create a hyperkube image and deployed the kubernetes cluster.

**Comment by divyenpatel** [comment]
@BaluDontu  to unblock your self for testing your changes. 
You can keep `mkfs.ext4` on local directory on the Node VM and mount that directory in Kubelet and add it to the Path in kubelet container, until we have fix available for https://github.com/kubernetes/kubernetes/issues/50802
you will need to do this on nodes in the cluster.

**Comment by divyenpatel** [comment]
New Ignition image is available at `cnastorage/k8s-ignition:v1.8-dev-release-with-kubelet-fix` to unblock testing.

Use this new image while deploying kubernetes-cluster. In this image kubelet container is configured to mount /usr/sbin from the docker host.

Added following volume in the kubelet container configuration
`-v /usr/sbin/:/usr/sbin/:ro \`

This makes mkfs.ext4 available in the kubelet running using the hyperkube image build from master branch.
